cluster-autoscaler/loop/trigger.go

```diff
@@ -43,38 +45,74 @@ type scalingTimesGetter interface {
+	provisioningRequestProcessTimeGetter provisioningRequestProcessingTimesGetter
 }
 
 // NewLoopTrigger creates a LoopTrigger object
-func NewLoopTrigger(podObserver *UnschedulablePodObserver, scalingTimesGetter scalingTimesGetter, scanInterval time.Duration) *LoopTrigger {
+func NewLoopTrigger(scalingTimesGetter scalingTimesGetter, provisioningRequestProcessTimeGetter provisioningRequestProcessingTimesGetter, scanInterval time.Duration) *LoopTrigger {
 	return &LoopTrigger{
-		podObserver:        podObserver,
-		scanInterval:       scanInterval,
-		scalingTimesGetter: scalingTimesGetter,
+		podObserver:                          nil,
+		scanInterval:                         scanInterval,
+		scalingTimesGetter:                   scalingTimesGetter,
+		provisioningRequestProcessTimeGetter: provisioningRequestProcessTimeGetter,
+	}
+}
+
+// Initialize initializes the LoopTrigger object by providing a poiumer to the UnschedulablePodObserver
+func (t *LoopTrigger) Initialize(podObserver *UnschedulablePodObserver) {
+	t.podObserver = podObserver
+	t.initialized = true
+}
+
+// provisioningRequestWasProcessed is used to check if provisioningRequestProcessTimeGetter is not nil and a provisioning request was processed in the last iteration
+func (t *LoopTrigger) provisioningRequestWasProcessed(lastRun time.Time) bool {
+	return t.provisioningRequestProcessTimeGetter != nil && !t.provisioningRequestProcessTimeGetter.LastProvisioningRequestProcessTime().Before(lastRun)
+}
+
+// logIfUnschedulablePodsDidNotAppear logs a message if the next iteration was not triggered by unschedulable pods appearing, else it logs a message that the next iteration was triggered by unschedulable pods appearing
+func (t *LoopTrigger) logIfUnschedulablePodsDidNotAppear(message string) {
+	select {
+	case <-t.podObserver.unschedulablePodChan:
+		klog.Info("Autoscaler loop triggered by unschedulable pod appearing")
+	default:
+		klog.Infof(message)
 	}
 }
 
 // Wait waits for the next autoscaling iteration
-func (t *LoopTrigger) Wait(lastRun time.Time) {
+func (t *LoopTrigger) Wait(lastRun time.Time) errors.AutoscalerError {
 	sleepStart := time.Now()
 	defer metrics.UpdateDurationFromStart(metrics.LoopWait, sleepStart)
 
 	// To improve scale-up throughput, Cluster Autoscaler starts new iteration
 	// immediately if the previous one was productive.
+	if !t.initialized {
+		return errors.ToAutoscalerError(errors.InternalError, fmt.Errorf("LoopTrigger not initialized"))
+	}
+
 	if !t.scalingTimesGetter.LastScaleUpTime().Before(lastRun) ||
 		!t.scalingTimesGetter.LastScaleDownDeleteTime().Before(lastRun) {
-		select {
-		case <-t.podObserver.unschedulablePodChan:
-			klog.Info("Autoscaler loop triggered by unschedulable pod appearing")
-		default:
-			klog.Infof("Autoscaler loop triggered immediately after a productive iteration")
-		}
-		return
+		t.logIfUnschedulablePodsDidNotAppear("Autoscaler loop triggered immediately after a productive iteration")
```

Comment:
I think it will be more readable if you rename the function to "triggerNextIteration".

"logIfUnschedulablePodsDidNotAppear" name makes sense, but it's confusing if you're passing a message unrelated to what function name says.

Reply:
Updated